File: app/src/icloud/java/at/bitfire/davdroid/ui/setup/ICloudLoginCredentialsFragment.java
package at.bitfire.davdroid.ui.setup;

import android.graphics.Typeface;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import java.net.URI;

import at.bitfire.davdroid.R;
import at.bitfire.davdroid.ui.widget.EditPassword;

public class ICloudLoginCredentialsFragment extends Fragment {

    EditText editUserName;
    EditPassword editUrlPassword;


    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.login_credentials_fragment, container, false);

        Typeface tf = Typeface.createFromAsset(getContext().getAssets(), "fonts/BebasNeue_Light.ttf");
        ((TextView)v.findViewById(R.id.login_title)).setTypeface(tf);

        editUserName = (EditText)v.findViewById(R.id.user_name);
        editUrlPassword = (EditPassword)v.findViewById(R.id.url_password);

        final Button login = (Button)v.findViewById(R.id.login);
        login.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                LoginCredentials credentials = validateLoginData();
                if (credentials != null)
                    DetectConfigurationFragment.newInstance(credentials).show(getFragmentManager(), null);
            }
        });

        return v;
    }

    protected LoginCredentials validateLoginData() {
        boolean valid = true;

        String userName = editUserName.getText().toString();
        if (userName.isEmpty()) {
            editUserName.setError(getString(R.string.login_user_name_required));
            valid = false;
        }

        String password = editUrlPassword.getText().toString();
        if (password.isEmpty()) {
            editUrlPassword.setError(getString(R.string.login_password_required));
            valid = false;
        }

        return valid ? new LoginCredentials(URI.create("https://icloud.com/"), userName, password) : null;
    }


    public static class Factory implements ILoginCredentialsFragment {

        @Override
        public Fragment getFragment() {
            return new ICloudLoginCredentialsFragment();
        }

    }

}

File: app/src/icloud/java/at/bitfire/davdroid/ui/setup/ICloudLoginCredentialsFragment.kt
package at.bitfire.davdroid.ui.setup;

import android.os.Bundle
import android.support.v4.app.Fragment
import android.text.Html
import android.text.method.LinkMovementMethod
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import at.bitfire.davdroid.R
import kotlinx.android.synthetic.icloud.login_credentials_fragment.view.*
import java.net.URI

class ICloudLoginCredentialsFragment: Fragment() {

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        val v = inflater.inflate(R.layout.login_credentials_fragment, container, false)

        v.login_password_hint.movementMethod = LinkMovementMethod.getInstance()
        v.login_password_hint.text = Html.fromHtml(getString(R.string.login_app_specific_password_hint))

        v.login.setOnClickListener {
            validateLoginData()?.let { info ->
                DetectConfigurationFragment.newInstance(info).show(fragmentManager, null)
            }
        }

        return v
    }

    private fun validateLoginData(): LoginInfo? {
        var valid = true

        val v = requireNotNull(view)
        val userName = v.user_name.text.toString()
        if (userName.isEmpty()) {
            v.user_name.error = getString(R.string.login_user_name_required)
            valid = false
        }

        val password = v.url_password.getText().toString()
        if (password.isEmpty()) {
            v.url_password.setError(getString(R.string.login_password_required))
            valid = false
        }

        return if (valid)
            LoginInfo(URI.create("https://icloud.com/"), userName, password)
        else
            null
    }


    class Factory: ILoginCredentialsFragment {
        override fun getFragment() = ICloudLoginCredentialsFragment()
    }

}
